src/main/java/bio/terra/service/upgrade/UpgradeService.java

```diff
@@ -0,0 +1,56 @@
+package bio.terra.service.upgrade;
+
+import bio.terra.common.exception.NotImplementedException;
+import bio.terra.model.UpgradeModel;
+import bio.terra.service.iam.AuthenticatedUserRequest;
+import bio.terra.service.job.JobService;
+import bio.terra.service.upgrade.exception.InvalidCustomNameException;
+import bio.terra.service.upgrade.flight.UpgradeProfileFlight;
+import bio.terra.stairway.Flight;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class UpgradeService {
+
+    private enum CustomFlight {
+        BILLING_PROFILE_PERMISSION(UpgradeProfileFlight.class);
+
+        private final Class<? extends Flight> flightClass;
+
+        CustomFlight(Class<? extends Flight> flightClass) {
+            this.flightClass = flightClass;
+        }
+
+        public Class<? extends Flight> getFlightClass() {
+            return flightClass;
+        }
+    }
+
+    private final JobService jobService;
+
+    @Autowired
+    public UpgradeService(JobService jobService) {
+        this.jobService = jobService;
+    }
+
+    public String upgrade(UpgradeModel request, AuthenticatedUserRequest user) {
+        if (request.getUpgradeType() != UpgradeModel.UpgradeTypeEnum.CUSTOM) {
+            throw new NotImplementedException("Upgrade type is not implemented: " + request.getUpgradeType().name());
+        }
+
+        CustomFlight customFlight;
+        try {
+            customFlight = CustomFlight.valueOf(request.getCustomName());
+        } catch (NullPointerException ex) {
```

Comment:
A nit but this strikes me as a strange way to do null checking...I'd expect that to be an explicit check right above.

Also, isn't that something that we could define in the openapi model spec?

Reply:
It is unusual. I was reacting to what gets thrown from valueOf.

The unreviewed design is that there will be more enumerations than just custom in the future, so I did not want to make it required in the openapi. For example, if we add liquibase operations, as I suggest, we would not require the custom name.

So I need to validate here. I could write code that does the check and loop itself and detects null. If I use valueOf, I'll have to deal with the NPE. So not sure what to change here.